Improved matrix power performance

Abdullahjavednesar opened the pull request:
Fixes  #11964

mamueller committed:
initial copy

mamueller committed:
changed names so that the new module is loaded and tested, although it still translates into c not R

mamueller committed:
renamed class

mamueller committed:
made the test for power expressions green in test_rcode1.py

mamueller committed:
all tests in test_rcode green

mamueller committed:
nested piecewise test now runs in test_rcode2.py

mamueller committed:
corrected some tabs and found a wrong test. There is a paranthesis
missing int piecewise expressions

mamueller committed:
added a wrong priecewise
test

mamueller committed:
repaired the piecewise test

mamueller committed:
tried to find out how the constants are handled and introduced a lot of
tracing that will be erased later

mamueller committed:
The tests for the constants now work
also removed the verbose debugging printouts

mamueller committed:
changed the translation for the for loops

mamueller committed:
Checked a little more of the translation. The next step is concerned
with arrays (matrices)

mamueller committed:
finished the translation of the tests of the ccode to the rcode module

mamueller committed:
corrected a comment that still refered to C as the target language

Commit:
Add info about GSoC 2015, 2016

smichr committed:
nsolve leaves function unmodified

Commit:
Add sigma approximation to FourierSeries class

Abdullahjavednesar commented:
@jksuom can you please review.
Thanks.

dhruvkb committed:
Add a second condition to the if statement to ensure that x is also an integer because the if condition only checked if the value of y was an integer

dhruvkb committed:
Add the case specified in bug #11955 to the test to make sure it stays fixed in future iterations

Abdullahjavednesar committed:
singular matrix fix

Abdullahjavednesar commented:
```
In [3]: Matrix([[a, 1, 0], [0, a, 1], [0, 0, a]])**n == Matrix([
   ...:         [a**n, a**(n-1)*n, a**(n-2)*(n-1)*n/2],
   ...:         [0, a**n, a**(n-1)*n],
   ...:         [0, 0, a**n]])
Out[3]: True
```
I'm unable to fix Job 20366.6. Can any one help.

Upabjojr commented on sympy/matrices/matrices.py:
```diff
@@ -192,6 +192,11 @@ def _matrix_pow_by_jordan_blocks(self, num):
 
+        l = list(sorted(self.eigenvals(rational=False).keys()))
+        for i in range(len(l)):
+            if l[i] == 0 and num < 0:
```
put this `if` on top, more efficient.

Upabjojr replied:
You can rewrite everything with a single `if` using `not any(eigenvals)`

Upabjojr replied:
Or maybe just use the determinant... way easier.

Abdullahjavednesar replied:
>put this if on top, more efficient.

do you mean before import..

Upabjojr commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -224,6 +224,9 @@ def test_power():
     raises(AttributeError, lambda: m._matrix_pow_by_jordan_blocks(10))
 
+    # test issue 11964
+    raises(ValueError, lambda: Matrix([[1, 1], [3, 3]])**10.0)
```
It's not a negative power. Call `_matrix_pow_by_jordan_blocks` instead of `**`

Abdullahjavednesar committed:
determinant used

Abdullahjavednesar commented:
hope this fixes all, ping @Upabjojr

Abdullahjavednesar committed:
efficient now

Upabjojr commented on sympy/matrices/matrices.py:
```diff
@@ -189,6 +189,9 @@ def __neg__(self):
         return -1*self
 
     def _matrix_pow_by_jordan_blocks(self, num):
+        if self.det() == 0 and num < 0:
```
Just one minor detail: `num < 0` before `self.det() == 0`, as it is faster to evaluated. This means the determinant will only be calculated if `num < 0`.

Upabjojr commented:
OK, just correct that minor detail, then it should be ready for merging.

Abdullahjavednesar committed:
Jordan Block improved

Abdullahjavednesar commented:
Can you merge @Upabjojr

Upabjojr commented:
Tests did not pass.

Upabjojr commented:
You have to take into account that `num` might be symbolic.

Maybe it's better to put that `if` in the upper function, that is in `Matrix.__pow__`.

Upabjojr commented:
@Abdullahjavednesar I got the impression that you are not properly debugging the code. Which debugger do you currently use?

Abdullahjavednesar commented:
Any suggestion?

Abdullahjavednesar committed:
tested OK

Abdullahjavednesar commented:
Ping @asmeurer @jksuom @Upabjojr can any one review.

mamueller committed:
Adapted the Rcode printer to the current version of sympy. All tests and doctests work

mamueller committed:
removed a superflous print statement that I had forgotten.

Upabjojr commented on sympy/matrices/matrices.py:
```diff
@@ -189,6 +189,8 @@ def __neg__(self):
         return -1*self
 
     def _matrix_pow_by_jordan_blocks(self, num):
+        if isinstance(num, (int, Integer, float)) and num < 0 and self.det() == 0:
```
have you tried `(num < 0) == True` ?

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -189,6 +189,9 @@ def __neg__(self):
         return -1*self
 
     def _matrix_pow_by_jordan_blocks(self, num):
+        if (num < 0) == True and self.det() == 0:
```
The determinant will be found when the Jordan blocks are computed. Would it be possible to use that instead of another computation here?

Upabjojr replied:
I'm not sure that `jordan_cells()` has a standardized output for blocks corresponding to zero eigenvalues:

```python

In [15]: m = Matrix([[1, 1, 1], [3, 3, 3], [2, 2, 2]])

In [16]: m
Out[16]: 
⎡1  1  1⎤
⎢       ⎥
⎢3  3  3⎥
⎢       ⎥
⎣2  2  2⎦

In [17]: m.jordan_cells()
Out[17]: 
⎛⎡-1  -1  1/2⎤, [[0], [0], [6]]⎞
⎜⎢           ⎥                 ⎟
⎜⎢1   0   3/2⎥                 ⎟
⎜⎢           ⎥                 ⎟
⎝⎣0   1    1 ⎦                 ⎠
```

In this case it's `[0], [0]`, but it could be
```
[[0, 0],
 [0, 0]]
```

jksuom replied:
The eigenvalues are considered [here](https://github.com/sympy/sympy/blob/master/sympy/matrices/matrices.py#L197). It would be possible to raise an exception at this point.

If the determinant is computed, I think it could be done before calling _matrix_pow_by_jordan_blocks.

Abdullahjavednesar replied:
>If the determinant is computed, I think it could be done before calling _matrix_pow_by_jordan_blocks

If det computed and raised in ```__pow__```, it won't work if we directly call ```_matrix_pow_by_jordan_blocks```.

mamueller committed:
Implemented the proposed changes to the comment on the pullrequest.

mamueller committed:
added the new parts to the documentation of the printing module

mamueller committed:
linked the rcode function to the codegen docs

Commit:
Fixed Typos

rajatthakur committed:
Fixes #11975

Upabjojr commented:
Calculating the determinant may cause some performance issues. @Abdullahjavednesar did you understand the proposal by @jksuom ? Do you feel like fixing it that way?

Abdullahjavednesar commented:
@jksuom I understand what you say and you're right as that's really gonna improve the performance, but my reply was about your second suggestion regarding determinant evaluation. I'll surely fix this using eigenvalues as that's better. I'll be committing  soon, actually I am travelling now.

dhruvkb committed:
Optimize calculations by creating a temporary variable temp. Also simplify the second expression based on input from @smichr

dhruvkb committed:
Add an small explanatory comment

moorepants committed:
Moves autowrap tests in with MATPLOTLIB/THEANO/ASCII tests.

- Uninstall cycle is now under the SAGE tests.
- autowrap related tests run with matplotlib, theano, an ascii tests.
- TEST_AUTOWRAP now uses conda installed packages.

moorepants committed:
Fixed yaml error.

moorepants committed:
Fix python 3 error and version comparison error.

Abdullahjavednesar committed:
eigen values used

Abdullahjavednesar commented:
Ping @jksuom

jksuom commented:
I think that only non-negative integer powers of singular matrices should be allowed. Otherwise the results are not useful.
```
In [1]: M = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 0]])

In [2]: M._matrix_pow_by_jordan_blocks(S(1)/2)
Out[2]: 
⎡nan  nan  nan⎤
⎢             ⎥
⎢nan  nan  nan⎥
⎢             ⎥
⎣nan  nan  nan⎦

In [3]: M._matrix_pow_by_jordan_blocks(S(3)/2)
Out[3]: 
⎡nan  nan  zoo⎤
⎢             ⎥
⎢nan  nan  nan⎥
⎢             ⎥
⎣nan  nan  nan⎦
```

Abdullahjavednesar commented:
@jksuom It seems than not all singular matrices behave that way.
```
In [17]: x
Out[17]:
Matrix([
[8, 8, 8],
[1, 1, 1],
[0, 0, 0]])

In [18]: x._matrix_pow_by_jordan_blocks(S(1)/2)
Out[18]:
Matrix([
[8/3, 8/3, 8/3],
[1/3, 1/3, 1/3],
[  0,   0,   0]])
```

jksuom commented:
>  It seems than not all singular matrices behave that way.

That is true. The result depends on the size of the Jordan blocks with eigenvalue 0 and on the exponent.

Abdullahjavednesar commented:
>The result depends on the size of the Jordan blocks with eigenvalue 0 and on the exponent.

I'm not sure, Is it the size of the matrix ```x.jordan_cells()[1]``` you're talking about?
If yes, how does it depends on its size?
I think the result remains useless if the exponent is float type and ```None ``` should be returned.
```
In [49]: x=Matrix([[0,1,0,0],[0,0,1,0],[0,0,0,1],[0,0,0,0]])

In [50]: x.jordan_cells()
Out[50]:
(Matrix([
 [1, 0, 0, 0],
 [0, 1, 0, 0],
 [0, 0, 1, 0],
 [0, 0, 0, 1]]), [Matrix([
  [0, 1, 0, 0],
  [0, 0, 1, 0],
  [0, 0, 0, 1],
  [0, 0, 0, 0]])])

In [51]: x.jordan_cells()[1]
Out[51]:
[Matrix([
 [0, 1, 0, 0],
 [0, 0, 1, 0],
 [0, 0, 0, 1],
 [0, 0, 0, 0]])]

```

Abdullahjavednesar commented:
@jksuom  can you help generalizing the size condition.

jksuom commented:
> How does it depends on its size of the Jordan blocks with eigenvalue 0?

A Jordan block with eigenvalue 0 is nilpotent but non-zero if its size is greater than 1. (Blocks of size 1 are equivalent to scalars and a zero block could probably be treated in the same way as powers of the scalar 0.) For a nilpotent matrix it is possible to define non-negative integral powers, and I think those should be the only exponents allowed in such a case.

> I think the result remains useless if the exponent is float type and None should be returned.

For a Jordan block with eigenvalue 0 and size > 1, float type exponents should probably not be allowed. But I think that an error should be raised instead of None.

jajodiaraghav committed:
powsimp() modified for noncommutative variables

jajodiaraghav committed:
modified powsimp() for noncommutative vars

jajodiaraghav committed:
modified powsimp() for noncommutative vars

Abdullahjavednesar committed:
singular matrix improved

Abdullahjavednesar committed:
corrected singular matrices

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1547,6 +1547,13 @@ def _matrix_pow_by_jordan_blocks(self, num):
             l = jc[0, 0]
+            if (num < 0) == True and l == 0:
+                raise ValueError("Matrix det == 0; not invertible")
+            if (num < 2) == True:
```
Where does 2 come from?

Abdullahjavednesar replied:
@jksuom this is what I observed hence concluded ```num < 2```
```
In [6]: x
Out[6]:
Matrix([
[0, 1, 0],
[0, 0, 1],
[0, 0, 0]])

In [7]: x._matrix_pow_by_jordan_blocks(1.9)
Out[7]:
Matrix([
[nan, nan, zoo],
[nan, nan, nan],
[nan, nan, nan]])

In [8]: x._matrix_pow_by_jordan_blocks(1.9999999999999999999999999)
Out[8]:
Matrix([
[0, 0, 0**0.0],
[0, 0,      0],
[0, 0,      0]])
```

jksuom replied:
But there are other matrices where 2 does not suffice, namely nilpotent Jordan blocks of size > 2.

Abdullahjavednesar replied:
Yes you're right, I think that should be corrected to 0 instead of 2

Abdullahjavednesar replied:
or should it be ```isinstance(num, float)``` but that won't include ```S(1)/2```, I guess.

jksuom replied:
I suggest that only non-negative integer exponents are allowed if there are nilpotent Jordan blocks (of size > 1). That could probably include floats that are equal to integers.

jajodiaraghav committed:
modified powsimp() for noncommutative vars

jajodiaraghav committed:
modified powsimp() for noncommutative vars

jajodiaraghav committed:
modified powsimp() for noncommutative vars

jajodiaraghav committed:
modified powsimp() for noncommutative vars

jajodiaraghav committed:
powsimp() modified for noncommutative variables

jajodiaraghav committed:
powsimp() modified for noncommutative variables

Abdullahjavednesar committed:
conditions corrected

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1549,11 +1549,12 @@ def jordan_cell_power(jc, n):
-                    if (b[i]).rows > 1 and b[i][0, 0] == 0:
-                        raise ValueError("Cannot be evaluated")
+            if not isinstance(num, (int, Integer, Expr)):
+                if not num - int(num) == 0:
```
It seems that this will never work if `num` is a symbol.

jksuom replied:
If `num` were a SymPy expression, it is would be possible to ask `if not num.is_integer`.
Since this is an internal method, the exponent could be sympified at entrance to `__pow__`.

Abdullahjavednesar replied:
Tried ```if not num.is_integer```, but throws Attribute error.
```
if not isinstance(num, (int, Integer)):
    if isinstance(num, float) and not num - int(num) == 0:
```
Is this better way to allow non negative integers and float which are integer exponents? Should ```Expr``` be allowed too?

jksuom replied:
`num.is_integer` will work only if `num` has been sympified. I think that should be done in `__pow__` *before* calling `_matrix_pow_by_jordan_blocks`. Then `num` is not an `int` or a `float`. A `num` of class `Expr` could probably be allowed, if `num.is_integer` is true.

jajodiaraghav committed:
powsimp() modified for noncommutative variables

bjodah committed:
Add kwarg 'ignore' to cse() for disqualification.

asmeurer committed:
Update citation information in the README

asmeurer committed:
Use > to block quote the paper citation

asmeurer committed:
Maybe just need to remove the ::

asmeurer committed:
Add citing information to the docs

asmeurer committed:
Move special topics to better place in the docs index

asmeurer committed:
Expose polys dup_transform as Poly.transform

This seems to be the first method on DMP that requires unifying three
polynomials. I hope I did the unification correctly. The tests that I've added
seem to indicate that I did.

This fixes #12003.

dhruvkb committed:
Rewrite the second condition so as to highlight the symmetry in the conditions of x and y while making the logic behind it clearer

asmeurer committed:
Don't use Unicode characters in the BibTeX citation for the SymPy paper

Unicode characters can cause issues in BibTeX. Use the LaTeX escapes instead.

c.f. #12014

CC @certik @rouckas

Abdullahjavednesar committed:
checked conditions

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1589,6 +1583,11 @@ def _matrix_pow_by_recursion(self, num):
     def __pow__(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        if not sympify(num).is_integer and (num > 0) == True and not num % 1 == 0:
```
It is better to start with `num = sympify(num)` before doing anything else.

Abdullahjavednesar replied:
Okay. Is the rest fine?

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1589,6 +1583,11 @@ def _matrix_pow_by_recursion(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        if not sympify(num).is_integer and (num > 0) == True and not num % 1 == 0:
+            b = self.jordan_cells()[1]
```
This call belongs to `matrix_pow_by_jordan_blocks`, not to `__pow__`.

Abdullahjavednesar replied:
Is the use ```self.jordan_form()``` good, or can you suggest a better alternative.

jksuom replied:
Computing the Jordan form is expensive. It should not be done at all when recursion is applied. Hence the role of `__pow__` is only to decide which method is to be used, Jordan forms or recursion. It can also sympify the exponent `num` so that tests like `num.is_integer` will work in `_matrix_power_by_jordan_blocks`.

mamueller committed:
Changed the few remaining forgotten references to C instead of R

debugger22 committed:
Update copyright year to the current year

gxyd committed:
Fix repr printing of Permutation and tests for str, repr printing

jksuom commented on sympy/concrete/logic/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .boolalg import (to_cnf, to_dnf, to_nnf, And, Or, Not, Xor, Nand, Nor, Implies,
```
How did logic appear in this PR?

Abdullahjavednesar replied:
I think something went wrong, I tried changing only matrices and the test file. @jksuom can you have a look on those, I'll remove this commit later.

Abdullahjavednesar committed:
corrected logic

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1583,11 +1583,11 @@ def _matrix_pow_by_recursion(self, num):
-                if (b[i]).rows > 1 and b[i][0, 0] == 0:
-                    raise ValueError("Cannot be evaluated")
+        num = sympify(num)
+        if not num.is_integer and (num >= 0) == True and not num % 1 == 0 and self.is_nilpotent():
```
It is seldom useful to ask if the whole matrix is nilpotent. It is only in the Jordan block method that it matters if a block is nilpotent. It should not be necessary to consider nilpotence yet at this stage.

Abdullahjavednesar replied:
considered nilpotence here as I'm extracting block size with eigenvalue 0 later ```z = self.eigenvals()[0]```

jksuom replied:
If the matrix is nilpotent, there is no need to call `self.eigenvals()` since 0 is the only eigenvalue.

Abdullahjavednesar replied:
Oops! it should be ```det()``` rather.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1581,6 +1583,11 @@ def _matrix_pow_by_recursion(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        num = sympify(num)
+        if not num.is_integer and (num >= 0) == True and not num % 1 == 0 and self.det() == 0:
```
The determinant is computed in connection with Jordan blocks in `_matrix_pow_by_jordan_blocks`. Computing it twice is not an improvement.

gxyd committed:
Update the GSoC info organising

asmeurer committed:
Fix pull request message to always be hidden

GitHub stopped automatically hiding it because the # character gave it away as
not a real HTML tag. Now mark it as an HTML comment, which should always be
hidden.

gxyd committed:
Change 2014 -> 2015

gxyd committed:
Add one more GSoC 2015 student in SciRuby section

bjodah committed:
Make test of cse with 'ignore' more terse

Abdullahjavednesar commented:
@jksuom, I think the logic is correct here, can you point out where am I going wrong?

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1581,6 +1583,12 @@ def _matrix_pow_by_recursion(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        num = sympify(num)
+        if not num.is_integer and (num >= 0) == True and not num % 1 == 0:
```
I don't understand these conditions. I think that `__pow__` should only decide which method to use. When the Jordan block method is used, it can find out whether the power can be evaluated or not after computing the blocks.

Abdullahjavednesar replied:
We just need to allow **non-negative integers** and **floats which are integers** and the rest if (```not num.is_integer and (num >= 0) == True and not num % 1 == 0```) the computed jordan block is nilpotent of size > 1 we raise an error. Here ```self.eigenvals()[0]``` computes the size of the block with eigenvalue 0. These are the conditions and logic applied, please correct me if I'm wrong anywhere or I'm missing something.

jksuom replied:
`self.eigenvals()[0]` computes the sum of the sizes of blocks with eigenvalue 0, not the size of a single block.

Abdullahjavednesar replied:
well, I had a doubt here thanks @jksuom I'll correct this. 
>I don't understand these conditions

Is the condition okay?

jksuom replied:
I wonder what should be done if `num < 0`. Also, what is the purpose of `not num % 1 == 0` if `not num.is_integer` is true?

Abdullahjavednesar replied:
if num < 0 under the condition when it has any eigenvalue 0 will return ```ValueError("Matrix det == 0; not invertible")```
```not num % 1 == 0``` to exclude float which are integers eg 10.0 as you suggested

jksuom replied:
Why cannot `num >= 0` be tested in the same place as `num < 0`? How does `not num % 1 == 0` exclude integer valued Floats?

Abdullahjavednesar replied:
```
In [1]: a=5

In [2]: b=5.0

In [3]: a%1
Out[3]: 0

In [4]: b%1
Out[4]: 0.0

In [5]: b%1==0
Out[5]: True
```

jksuom replied:
Is it necessary to have both `not num.is_integer` and `not num % 1 == 0`? It looks like one condition would imply the other.

Abdullahjavednesar replied:
I don't find them same?
```
In [9]: sympify(a).is_integer
Out[9]: False

In [10]: sympify(a)%1==0
Out[10]: True

In [11]: a
Out[11]: 10.0
```

jksuom replied:
The conditions are obviously not equivalent. But is is necessary to have both conditions? If `num %  1` is different from zero, can `num.is_integer` still be true?

Abdullahjavednesar replied:
I think in that case ```num % 1``` includes ```num.is_integer``` and is sufficient.

Abdullahjavednesar replied:
In sympy\physics\quantum\density.py
```
>>> fidelity(up_dm, down_dm) #orthogonal states
    0
```
I think this is going to fail under the new conditions (has nilpotent jordan block with size > 1)

gxyd committed:
Change number of PSF 2015 GSoC students for SymPy

tyagi-iiitv committed:
Removed the redundant import

sympy.assumptions.ask import Q was called twice.

Abdullahjavednesar commented:
@jksuom Is this ready? can you please have a look.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1594,6 +1596,12 @@ def __pow__(self, num):
         elif isinstance(num, (Expr, float)):
+            num = sympify(num)
+            if (num >= 0) == True and not num % 1 == 0:
+                z = self.jordan_form()[1]
```
`jordan_form()` calls `jordan_cells()` that is also called by `_matrix_pow_by_jordan_blocks()`. It is fairly expensive, and calling it twice should be avoided.

Abdullahjavednesar replied:
Can you suggest an alternative?

jksuom replied:
There should be no need for an alternative when the Jordan block method is used. In case recursion is used, and the exponent is negative, trying to compute the inverse will reveal singular matrices.

Abdullahjavednesar replied:
But we'd need to test the nilpotence of a block with size > 1, where should that be done when the exponent is not a non negative integer? Can you please elaborate.

jksuom replied:
I think that should be done in `_matrix_pow_by_jordan_blocks` after the Jordan cells have been found. That way there would be no duplicate computation.

Abdullahjavednesar replied:
```
 if (num >= 0) == True and not num % 1 == 0 and jc.rows > 1 and l == 0: 
 ```
So this condition must be sufficient in ```_matrix_pow_by_jordan_blocks``` to raise an error. After this I think there's no need to change anything in```__pow__```.

jksuom replied:
The only thing that I would consider changing in `__pow__` is sympifiying `num`. If that is done, it is not necessary to ask if `num` is an `int` or  a `float`. Instead, the tests could be `num.is_Integer` and `isinstance(num, Expr)`. Also, the tests in `_matrix_pow_by_jordan_blocks` could rely on `num` being sympified. But maybe that is not important and not necessary now.

As to the conditions in `jordan_cell_power`, I would take them in different order, first if `l == 0` and `jc.rows > 1`, and only then ask if `n` is a non-negative integer or not. That way, only one condition (`l == 0`) would have to be asked in most cases.

Abdullahjavednesar replied:
>If that is done, it is not necessary to ask if ```num``` is an ```int``` or a ```float```

In case of ```float``` will it help? But yes ```isinstance(num, (int, Integer))``` can be replaced.
```
In [13]: not S(a).is_integer
Out[13]: True

In [14]: not S(50.2).is_integer
Out[14]: True

In [15]: S(a).is_integer
```

jksuom replied:
I thought that `num.is_Integer` (not `num.is_integer`) could be used when we know that `num` cannot be an int. Currently all floats are handled by the Jordan block method, and that could also be the case after they had been sympified to Floats (which satisfy the condition `isinstance(num, Expr)`). But it may be worth investigating if the recursion method could be used with integer valued floats like 50.0 (in which case the first condition would have to be modified accordingly).

Abdullahjavednesar replied:
Integer valued floats can be calculated using recursion and a benefit would be a simplified result. Rest floats can only be computed using jordan block method.

Abdullahjavednesar committed:
added related tests

Abdullahjavednesar commented:
Instead ``` if num.is_Integer```, ```if num % 1 == 0```would be just sufficient to include integer valued floats in ```__pow__```.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1547,6 +1547,10 @@ def _matrix_pow_by_jordan_blocks(self, num):
         def jordan_cell_power(jc, n):
             N = jc.shape[0]
             l = jc[0, 0]
+            if l == 0 and (num < 0) == True:
```
Perhaps this should contain the argument `n` instead of the global `num`.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1547,6 +1547,10 @@ def _matrix_pow_by_jordan_blocks(self, num):
+            if l == 0 and (num < 0) == True:
+                raise ValueError("Matrix det == 0; not invertible")
+            elif l == 0 and jc.rows > 1 and (n >= 0) == True and not n % 1 == 0:
+                raise ValueError("Cannot be evaluated")
```
The explanation could be something like "Non-integer power cannot be evaluated".

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1547,6 +1547,10 @@ def _matrix_pow_by_jordan_blocks(self, num):
             l = jc[0, 0]
+            if l == 0 and (num < 0) == True:
+                raise ValueError("Matrix det == 0; not invertible")
+            elif l == 0 and jc.rows > 1 and (n >= 0) == True and not n % 1 == 0:
```
Is `(n >= 0) == True` necessary when `n < 0` is checked above?
`not n % 1 == 0` can be written `n % 1 != 0`.

jksuom commented:
Yes, I think `num % 1 == 0` could be a useful alternative and should be investigated. (I did forget to answer.)
There should also be a sufficient number of tests covering different types of matrices and exponents. I'm not sure if the existing ones would suffice.

Abdullahjavednesar commented:
Can you suggest some tests you'd like to be taken under consideration.

jksuom commented:
In principle, all combinations should be tested (singular and nonsingular matrices, different kinds of exponents). It should be checked if something is missing.

Abdullahjavednesar commented:
Are these tests sufficient?
```
    # test issue 11964
    raises(ValueError, lambda: Matrix([[1, 1], [3, 3]])._matrix_pow_by_jordan_blocks(-10))
    x = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 0]])
    assert x**10.0 == Matrix([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
    raises(ValueError, lambda: x**2.1)
    raises(ValueError, lambda: x**(S(3)/2))
    x = Matrix([[8, 1], [3, 2]])
    assert x**10.0 == Matrix([[1760744107, 272388050], [817164150, 126415807]])
    x = Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
    assert x**10.2 == Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
    x = Matrix([[0, 0, 1], [3, 0, 1], [4, 3, 1]])
    assert x**5.0 == Matrix([[168,  72,  89], [291, 144, 161], [572, 267, 329]])
    assert x**5.0 == x**5
```

jksuom commented:
This looks like a good collection. There could be singular matrices having Jordan blocks of different sizes. (Matrices could also be denoted by capital letters following the common convention.)

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1593,7 +1598,7 @@ def __pow__(self, num):
                     return self._matrix_pow_by_recursion(num)
             return self._matrix_pow_by_recursion(num)
-        elif isinstance(num, (Expr, float)):
+        elif num.is_Float or isinstance(num, Expr):
```
It is not necessary check for Floats since they are expressions.
```
In [5]: isinstance(Float(5.0), Expr)
Out[5]: True
```

jksuom commented:
Symbolic exponents should also be tested. `n = Symbol('n', integer=True)`, `n = Symbol('n', integer=True, nonnegative=True)`.

Abdullahjavednesar commented:
In ```__pow__``` ```num%1 == 0``` is not just sufficient ```num.is_Integer``` should also be included.
```
In [2]: n = Symbol('n', integer=True)

In [3]: x=Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])

In [4]: x**n
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-bc13e94c104e> in <module>()
----> 1 x**n

C:\Users\user\sympy\sympy\core\decorators.py in binary_op_wrapper(self, other)
    116                     else:
    117                         return f(self)
--> 118             return func(self, other)
    119         return binary_op_wrapper
    120     return priority_decorator

C:\Users\user\sympy\sympy\matrices\dense.py in __pow__(self, other)
    137     @call_highest_priority('__rpow__')
    138     def __pow__(self, other):
--> 139         return super(DenseMatrix, self).__pow__(other)
    140
    141     @call_highest_priority('__add__')

C:\Users\user\sympy\sympy\matrices\matrices.py in __pow__(self, num)
   1598                 except AttributeError:
   1599                     return self._matrix_pow_by_recursion(num)
-> 1600             return self._matrix_pow_by_recursion(num)
   1601         elif num.is_Float or isinstance(num, Expr):
   1602             return self._matrix_pow_by_jordan_blocks(num)

C:\Users\user\sympy\sympy\matrices\matrices.py in _matrix_pow_by_recursion(self,
 num)
   1568     def _matrix_pow_by_recursion(self, num):
   1569         from sympy.matrices import eye
-> 1570         n = int(num)
   1571         if n < 0:
   1572             return self.inv()**-n   # A**-2 = (A**-1)**2

C:\Users\user\sympy\sympy\core\expr.py in __int__(self)
    206         from sympy import Dummy
    207         if not self.is_number:
--> 208             raise TypeError("can't convert symbols to int")
    209         r = self.round(2)
    210         if not r.is_Number:

TypeError: can't convert symbols to int
```
and the reason is here```n%1 == 0``` is True.

jksuom commented:
Recursion should be attempted only when the exponent is a number, so a condition (such as `num.is_number` or `num.is_Number`) should be added even before `num % 1 == 0` is checked.

Abdullahjavednesar committed:
exlanatory comment

Abdullahjavednesar commented:
sympy/simplify/tests/test_cse.py fails in Python 3.5, unable to figure out why?

jksuom commented:
That is caused by a random bug in cse. I've restarted the job.

jksuom commented:
The tests with exponent n suggest that some conditions should be sharpened. There appear powers like `0**n` and `0**(n - 1)` in the matrices. They may lead to errors, if one wants to replace n with a non-positive integer. I'm not sure how that should be solved, but it seems that at least the condition `(n < 0) == True` should be replaced by one that will catch exponents like `n = Symbol('n', integer=True)` which may take negative values.

Abdullahjavednesar commented:
Hopefully ```ask(Q.negative(n)) == None or l == 0 and n < 0``` would be more efficient than ```l == 0 and (n < 0) == True```

Abdullahjavednesar commented:
hopefully ```if l == 0 and ( ask(Q.negative(n)) == None or n < 0)``` would be a better condition.

jksuom commented:
Maybe a single condition would suffice instead of both `Q.negative(n)` and `n < 0`. How would `(n < 0) != False` succeed?

Abdullahjavednesar commented:
```n < 0 != False``` would include ```n < 0 == True``` and ```n < 0 == None``` so won't these conditions (```if l == 0 and ( ask(Q.negative(n)) == None or n < 0)```) be needed?

jksuom commented:
I did not mean that `l == 0` should be omitted. That should remain, of course. I was concerned about the possible negative sign of `n`. Why should there be both `Q.negative(n)` and `n < 0` which should mean the same thing?

Abdullahjavednesar commented:
I didn't mean ```l == 0```, I think in cases where we are not sure about the number for example,
```
In [12]: n = Symbol('n', integer=True)
In [13]: ask(Q.negative(n))
In [14]: ask(Q.nonnegative(n))
In [15]: ask(Q.nonpositive(n))
In [16]: ask(Q.positive(n))
```
under such cases ```ask(Q.negative(n)) == None``` may help but ```n < 0``` won't.

jksuom commented:
It seems to me that `n < 0` could be used:
```
In [1]: (n < 0) != False
Out[1]: True
```

jksuom commented:
This could also work:
```
In [2]: ask(Q.negative(n)) != False
Out[2]: True
```
but it takes more time.

Abdullahjavednesar commented:
Yes, @jksuom you are right. I've changed as you suggested.

jksuom commented:
There could still be one issue with `n = Symbol('n', integer=True, nonnegative=True)`. If there is a nilpotent Jordan block of size > 1, there appear powers like `0**(n - 1)` that raise error if `n = 0` is substituted. They come from `jc[j, i+j] = l**(n-i)*bn` where `i in range(N)`. To guarantee that the exponent `n - i` is not negative in case `l === 0` we should  take care that `n` is not less than the largest value of `i`, namely `N - 1`. That could probably be handled by using `(n < N - 1) != False` instead of `(n < 0) != False`. This could be tested with expressions like `A**(n + 2)`.

It also seems to me that `jc.rows` is the same as the local variable `N`. Hence `N > 1` could be used instead of `jc.rows > 1`.

Abdullahjavednesar commented:
@jksuom Can you please restart the job.

jksuom commented:
Restart does not seem to help. I don't understand why all Python 3.5 jobs fail. Hopefully that will be cleared up. Meanwhile, you could consider adding some explanatory comments concerning Floats to the conditions in `__pow__` to help maintain the code. Namely, how integer valued Floats can be used by the recursive method and others can only be handled by the Jordan block method.

Abdullahjavednesar commented:
@jksuom I've added comment lines. I think now it is ready once Python 3.5 jobs pass.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1581,7 +1585,12 @@ def _matrix_pow_by_recursion(self, num):
             raise NonSquareMatrixError()
-        if isinstance(num, (int, Integer)):
+        num = sympify(num)
+        if num.is_Number and num % 1 == 0:
```
[PEP-8](https://www.python.org/dev/peps/pep-0008/#id30) says
> Block comments generally apply to some (or all) code that follows them, and are indented to the same level as that code.

I believe this means that this if-line should come *after* the block comment.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1581,7 +1585,12 @@ def _matrix_pow_by_recursion(self, num):
+        if num.is_Number and num % 1 == 0:
+        # Conditions to include Integers and Integer valued Floats (eg: 10.0, 26.0, etc.,).
+        # n = int(num) in _matrix_pow_by_recursion() can only work correctly
+        # with those Floats which are Integers.
```
I would suggest: 'with integer valued numbers.'

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -1581,7 +1585,12 @@ def _matrix_pow_by_recursion(self, num):
+        # Conditions to include Integers and Integer valued Floats (eg: 10.0, 26.0, etc.,).
+        # n = int(num) in _matrix_pow_by_recursion() can only work correctly
+        # with those Floats which are Integers.
+        # Rest Floats are thus handled by _matrix_pow_by_jordan_blocks().
```
Perhaps Rest -> Other

jksuom commented:
I added some comments on style, but otherwise I think the code could be merged. Only the issue with Python 3.5 should be resolved first.

Abdullahjavednesar committed:
comment before condition

madan96 committed:
Update .travis.yml

madan96 committed:
Install mpmath using pip

gxyd committed:
Fix Subs._eval_derivative function for higher order derivative

Fixes #12005, one of the test cases(only one) taken from diofant

asmeurer committed:
Add an nsolve test that would fail if nsolve only used the numerator

baoqchau committed:
updating gmpy with gmpy2 version 2.0.8 in Travis CI build

gxyd committed:
Replace derivative of expr with 0, when s is variable
if s is one of vars, the derivative of expr with respect to s
should not be included.

gxyd committed:
Use Add class as common for two terms, add one more tests

Idea for common class from diofant/diofant 3f7ca353 commit

barun511 committed:
Fix #12074 : test_latex throws an exception

jksuom commented:
@Abdullahjavednesar The number of commits could be reduced by squashing some of those that essentially cancel each other, if you can do that. Afterwards, the current master should be merged to make the Python 3.5 tests pass.

Abdullahjavednesar commented:
@jksuom Although I managed to resolve the problem I asked in gitter, but could not successfully squash commits in git bash, it's a request if you could squash them. Thanks!

Abdullahjavednesar commented:
Oops! squashing messed up I guess, I'm closing this and opening a new PR.
